onnxruntime/core/optimizer/free_dim_override_transformer.cc

```diff
@@ -90,6 +90,11 @@ Status FreeDimensionOverrideTransformer::ApplyImpl(Graph& graph, bool& modified,
         } else {
           // Set the dimension override
           new_dimension->set_dim_value(dimension_override);
+
+          if (graph.IsInitializedTensor(graph_input->Name())) {
+            graph.OverrideInitializedTensorDim(graph_input->Name(), dim_index, dimension_override);
+          }
```

Comment:
Can you clarify why the existing setup doesn't work? 

It looks like the existing free_dim_override_transformer code updates the shape in the NodeArg for the value to make a previously dynamic dimension fixed. In graph.cc we check that the shape from the NodeArg is compatible with the fixed shape of the initializer. As the initializer in this case is providing a default value in case the graph input is not provided, it must be compatible with the shape in the NodeArg for the graph input. 

It seems like you're creating a setup where you need/want to change a fixed dimension of the initializer to a different value but I don't understand how that is possible. If say I have an initializer with shape {2,4} it has 8 elements. What is meant to happen if you say the shape is now {4,4}? or {1,4}?

Reply:
The current setup doesn't work because it assumes that the overridden shape must be the same as the initializer's shape, which seems like an obvious assumption a priori but that doesn't always need to be true.

There are many operators where the output shape isn't solely determined by the input's shape (e.g. Reduce, MatMul). This is essentially the entire goal of LoRA weights: changing a subgraph of the model by using tensors of different ranks, because the output of the MatMul doesn't change the shape inference of the rest of the graph. We already allow this behavior without dimension override by allowing inputs to have free dimensions that are different from the initializers that they override, so all this change does is enabling it the same behavior for dimension overriding.

If it makes it better and if it's possible, I'd be open to add a shape inference step after this transformer to make sure that the shape overrides result in a valid graph.

Reply:
Maybe a concrete example would help. Graph input shape, initializer shape, change being made by free dim override. I want to understand how the relationship between a dynamic dimension in the graph input shape and the fixed value for that dim in the initializer is changed. 

Is this only allowed to affect dim values where the initializer has a value of 1 or some multiple of the new value so broadcasting is possible?